statsmodels/censored/robustros.py

```diff
@@ -0,0 +1,462 @@
+
+    return nondetects.append(detects)
+
+class RobustROSEstimator(object):
+    '''
+    Class to implement the Robust regression-on-order statistics (ROS)
+    method outlined in Nondetects and Data Analysis by Dennis R. Helsel
+    (2005) to estimate the left-censored (non-detect) values of a
+    dataset.
+
+    Parameters
+    ----------
+    data : pandas DataFrame
+        The censored dataset for which the non-detect values need to be
+        estimated.
+
+    result_col : optional string (default='res')
+        The name of the column containing the numerical values of the
+        dataset. Left-censored values should be set to the detection
+        limit.
+
+    censorship_col : optional string (default='cen')
+        The name of the column containing indicating which observations
+        are censored.
+        `True` implies Left-censorship. `False` -> uncensored.
+
+    transform_in : optional function or None (default = np.log)
+        By default, lognormality is assumed and the linear regression
+        is performed on log-transformed data. An arbitrary transformation
+        function can be passed using this parameter. If `None` is passed,
+        the data will remain untransformed.
+
+    transform_out : optional function or None (default = np.exp)
+        This is used after the linear regression to revert the trasformation
+        performed by `transform_in`. If `None` is passed, the data will remain
+        untransformed.
+
+    Attributes
+    ----------
+    N_obs : int
+        Total number of results in the dataset
+
+    N_cen : int
+        Total number of non-detect results in the dataset.
+
+    cohn : pandas DataFrame
+        A DataFrame of the unique detection limits found in `data` along
+        with the `A`, `B`, `C`, and `PE` quantities computed by the
+        estimation.
+
+    data : pandas DataFrame
+        An expanded version of the original dataset `data` passed the
+        constructor included in the `modeled` column.
+
+    debug : pandas DataFrame
+        A full version of the `data` DataFrame that includes preliminary
+        quantities.
+
+    Example
+    -------
+    >>> from lifelines.estimation import RobustROS
+    >>> ros = RobustROSFitter(myDataFrame, result_col='conc',
+                              censorship_col='censored')
+    >>> ros.fit()
+    >>> print(ros.data)
+
+    Notes
+    -----
+    It is inappropriate to replace specific left-censored values with
+    the estimated values from this method. The estimated values
+    (self.data['modeled']) should instead be used to refine descriptive
+    statistics of the dataset as a whole.
+
+    '''
+    def __init__(self, data, result_col='res', censorship_col='cen',
+                 transform_in=np.log, transform_out=np.exp):
+
+        # confirm a datatype real quick
+
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError("Input `data` must be a pandas DataFrame")
```

Comment:
Excellent point. I think we could easily handle three cases:

 1. `data` is a DataFrame and the column names are specified
 2. `data` is an array and the column indexes are specified
 3. `data` is `None` and 1-D arrays are fed into `result_col` and `censorship_col`